File: wagerproof-ios-native/WagerproofKit/Sources/WagerproofModels/EditorPick.swift
import Foundation

/// Codable mirror of the RN `EditorPick` type in
/// `wagerproof-mobile/types/editorsPicks.ts`. Field names match the RN
/// `editors_picks` table byte-for-byte so the same JSON payload from
/// main Supabase decodes cleanly here.
///
/// `archived_game_data` is loosely-typed in RN (`any`). We model it as a
/// concrete `ArchivedGameData` struct so the Swift side can read the same
/// keys both camelCased (`awayTeam`, `awaySpread`) and snake_cased
/// (`away_team`, `away_spread`) — RN historically writes both styles, so the
/// decoder accepts either.
public struct EditorPick: Codable, Identifiable, Sendable, Hashable {
    public let id: String
    public let gameId: String
    public let gameType: GameType
    public let editorId: String
    public let selectedBetType: String
    public let editorsNotes: String?
    public let isPublished: Bool
    public let createdAt: String
    public let updatedAt: String
    public let betslipLinks: [String: String]?
    public let pickValue: String?
    public let bestPrice: String?
    public let sportsbook: String?
    public let units: Double?
    public let isFreePick: Bool?
    public let archivedGameData: ArchivedGameData?
    public let betType: String?
    public let result: PickResult?

    public init(
        id: String,
        gameId: String,
        gameType: GameType,
        editorId: String,
        selectedBetType: String,
        editorsNotes: String? = nil,
        isPublished: Bool,
        createdAt: String,
        updatedAt: String,
        betslipLinks: [String: String]? = nil,
        pickValue: String? = nil,
        bestPrice: String? = nil,
        sportsbook: String? = nil,
        units: Double? = nil,
        isFreePick: Bool? = nil,
        archivedGameData: ArchivedGameData? = nil,
        betType: String? = nil,
        result: PickResult? = nil
    ) {
        self.id = id
        self.gameId = gameId
        self.gameType = gameType
        self.editorId = editorId
        self.selectedBetType = selectedBetType
        self.editorsNotes = editorsNotes
        self.isPublished = isPublished
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.betslipLinks = betslipLinks
        self.pickValue = pickValue
        self.bestPrice = bestPrice
        self.sportsbook = sportsbook
        self.units = units
        self.isFreePick = isFreePick
        self.archivedGameData = archivedGameData
        self.betType = betType
        self.result = result
    }

    enum CodingKeys: String, CodingKey {
        case id
        case gameId = "game_id"
        case gameType = "game_type"
        case editorId = "editor_id"
        case selectedBetType = "selected_bet_type"
        case editorsNotes = "editors_notes"
        case isPublished = "is_published"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case betslipLinks = "betslip_links"
        case pickValue = "pick_value"
        case bestPrice = "best_price"
        case sportsbook
        case units
        case isFreePick = "is_free_pick"
        case archivedGameData = "archived_game_data"
        case betType = "bet_type"
        case result
    }
}

/// Sport keys used by `editors_picks.game_type`. Maps 1:1 to the RN union
/// `'nfl' | 'cfb' | 'nba' | 'ncaab'`. Unknown values decode to `.unknown`
/// so a stray row doesn't blow up the whole feed.
public enum GameType: String, Codable, Sendable, Hashable, CaseIterable {
    case nfl
    case cfb
    case nba
    case ncaab
    case mlb
    case unknown

    public init(from decoder: Decoder) throws {
        let raw = (try? decoder.singleValueContainer().decode(String.self))?.lowercased() ?? ""
        self = GameType(rawValue: raw) ?? .unknown
    }

    public var displayLabel: String {
        switch self {
        case .nfl: return "NFL"
        case .cfb: return "CFB"
        case .nba: return "NBA"
        case .ncaab: return "NCAAB"
        case .mlb: return "MLB"
        case .unknown: return "—"
        }
    }
}

/// `result` column on `editors_picks`. Mirrors RN union exactly.
public enum PickResult: String, Codable, Sendable, Hashable {
    case won
    case lost
    case push
    case pending
}

/// Sub-doc inside `editors_picks.archived_game_data`. RN writes this JSON
/// blob when an editor saves a pick — it snapshots the betting lines so the
/// pick still renders correctly after the live game data scrolls off the
/// CFB Supabase tables. RN's persistence is inconsistent — some keys are
/// camelCased (`awayTeam`), others snake_cased (`away_team`) — so this
/// decoder accepts both. Encoding always uses camelCase (RN's newer format).
public struct ArchivedGameData: Codable, Sendable, Hashable {
    public var awayTeam: String?
    public var homeTeam: String?
    public var awayLogo: String?
    public var homeLogo: String?
    public var gameDate: String?
    public var gameTime: String?
    public var rawGameDate: String?
    public var awaySpread: Double?
    public var homeSpread: Double?
    public var overLine: Double?
    public var awayMl: Int?
    public var homeMl: Int?
    public var awayTeamColors: TeamColors?
    public var homeTeamColors: TeamColors?

    public init(
        awayTeam: String? = nil,
        homeTeam: String? = nil,
        awayLogo: String? = nil,
        homeLogo: String? = nil,
        gameDate: String? = nil,
        gameTime: String? = nil,
        rawGameDate: String? = nil,
        awaySpread: Double? = nil,
        homeSpread: Double? = nil,
        overLine: Double? = nil,
        awayMl: Int? = nil,
        homeMl: Int? = nil,
        awayTeamColors: TeamColors? = nil,
        homeTeamColors: TeamColors? = nil
    ) {
        self.awayTeam = awayTeam
        self.homeTeam = homeTeam
        self.awayLogo = awayLogo
        self.homeLogo = homeLogo
        self.gameDate = gameDate
        self.gameTime = gameTime
        self.rawGameDate = rawGameDate
        self.awaySpread = awaySpread
        self.homeSpread = homeSpread
        self.overLine = overLine
        self.awayMl = awayMl
        self.homeMl = homeMl
        self.awayTeamColors = awayTeamColors
        self.homeTeamColors = homeTeamColors
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: DynamicKey.self)
        // Accept either casing for each field.
        awayTeam = c.firstString(["awayTeam", "away_team"])
        homeTeam = c.firstString(["homeTeam", "home_team"])
        awayLogo = c.firstString(["awayLogo", "away_logo"])
        homeLogo = c.firstString(["homeLogo", "home_logo"])
        gameDate = c.firstString(["gameDate", "game_date"])
        gameTime = c.firstString(["gameTime", "game_time"])
        rawGameDate = c.firstString(["rawGameDate", "raw_game_date"])
        awaySpread = c.firstDouble(["awaySpread", "away_spread"])
        homeSpread = c.firstDouble(["homeSpread", "home_spread"])
        overLine = c.firstDouble(["overLine", "over_line"])
        awayMl = c.firstInt(["awayMl", "away_ml"])
        homeMl = c.firstInt(["homeMl", "home_ml"])
        awayTeamColors = c.firstDecodable(["awayTeamColors", "away_team_colors"])
        homeTeamColors = c.firstDecodable(["homeTeamColors", "home_team_colors"])
    }

    public func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: DynamicKey.self)
        try c.encodeIfPresent(awayTeam, forKey: DynamicKey(string: "awayTeam"))
        try c.encodeIfPresent(homeTeam, forKey: DynamicKey(string: "homeTeam"))
        try c.encodeIfPresent(awayLogo, forKey: DynamicKey(string: "awayLogo"))
        try c.encodeIfPresent(homeLogo, forKey: DynamicKey(string: "homeLogo"))
        try c.encodeIfPresent(gameDate, forKey: DynamicKey(string: "gameDate"))
        try c.encodeIfPresent(gameTime, forKey: DynamicKey(string: "gameTime"))
        try c.encodeIfPresent(rawGameDate, forKey: DynamicKey(string: "rawGameDate"))
        try c.encodeIfPresent(awaySpread, forKey: DynamicKey(string: "awaySpread"))
        try c.encodeIfPresent(homeSpread, forKey: DynamicKey(string: "homeSpread"))
        try c.encodeIfPresent(overLine, forKey: DynamicKey(string: "overLine"))
        try c.encodeIfPresent(awayMl, forKey: DynamicKey(string: "awayMl"))
        try c.encodeIfPresent(homeMl, forKey: DynamicKey(string: "homeMl"))
        try c.encodeIfPresent(awayTeamColors, forKey: DynamicKey(string: "awayTeamColors"))
        try c.encodeIfPresent(homeTeamColors, forKey: DynamicKey(string: "homeTeamColors"))
    }
}

public struct TeamColors: Codable, Sendable, Hashable {
    public let primary: String
    public let secondary: String

    public init(primary: String, secondary: String) {
        self.primary = primary
        self.secondary = secondary
    }

    public static let `default` = TeamColors(primary: "#6B7280", secondary: "#9CA3AF")
}

/// Per-pick game context joined from the relevant sport's input/lines table
/// (NFL, CFB, NBA, NCAAB, MLB). Mirrors RN's `GameData` type but Swiftified
/// — colors come back as `TeamColors`, optional values stay optional.
public struct EditorPickGameData: Sendable, Hashable {
    public var awayTeam: String
    public var homeTeam: String
    public var awayLogo: String?
    public var homeLogo: String?
    public var gameDate: String?
    public var gameTime: String?
    /// ISO-ish raw date string used for sorting / filtering. May be `YYYY-MM-DD`
    /// or full ISO 8601 — mirrors the RN code's mixed input.
    public var rawGameDate: String?
    public var awaySpread: Double?
    public var homeSpread: Double?
    public var overLine: Double?
    public var awayMl: Int?
    public var homeMl: Int?
    public var openingSpread: Double?
    public var awayTeamColors: TeamColors
    public var homeTeamColors: TeamColors

    public init(
        awayTeam: String,
        homeTeam: String,
        awayLogo: String? = nil,
        homeLogo: String? = nil,
        gameDate: String? = nil,
        gameTime: String? = nil,
        rawGameDate: String? = nil,
        awaySpread: Double? = nil,
        homeSpread: Double? = nil,
        overLine: Double? = nil,
        awayMl: Int? = nil,
        homeMl: Int? = nil,
        openingSpread: Double? = nil,
        awayTeamColors: TeamColors = .default,
        homeTeamColors: TeamColors = .default
    ) {
        self.awayTeam = awayTeam
        self.homeTeam = homeTeam
        self.awayLogo = awayLogo
        self.homeLogo = homeLogo
        self.gameDate = gameDate
        self.gameTime = gameTime
        self.rawGameDate = rawGameDate
        self.awaySpread = awaySpread
        self.homeSpread = homeSpread
        self.overLine = overLine
        self.awayMl = awayMl
        self.homeMl = homeMl
        self.openingSpread = openingSpread
        self.awayTeamColors = awayTeamColors
        self.homeTeamColors = homeTeamColors
    }
}

// MARK: - Codable helpers

/// Dynamic CodingKey used to accept either snake_case or camelCase JSON keys
/// without writing two parallel decoders for each.
struct DynamicKey: CodingKey {
    var stringValue: String
    var intValue: Int? { nil }
    init(string: String) { stringValue = string }
    init?(stringValue: String) { self.stringValue = stringValue }
    init?(intValue: Int) { return nil }
}

private extension KeyedDecodingContainer where Key == DynamicKey {
    // Swift's `try?` flattens `T?` from a throwing function — so a single
    // `if let v = try? decodeIfPresent(...)` already returns the unwrapped
    // value. No need for `v != nil` (would always be true).
    func firstString(_ keys: [String]) -> String? {
        for key in keys {
            if let dk = DynamicKey(stringValue: key),
               let v = try? decodeIfPresent(String.self, forKey: dk) {
                return v
            }
        }
        return nil
    }
    func firstDouble(_ keys: [String]) -> Double? {
        for key in keys {
            if let dk = DynamicKey(stringValue: key),
               let v = try? decodeIfPresent(Double.self, forKey: dk) {
                return v
            }
        }
        return nil
    }
    func firstInt(_ keys: [String]) -> Int? {
        for key in keys {
            if let dk = DynamicKey(stringValue: key),
               let v = try? decodeIfPresent(Int.self, forKey: dk) {
                return v
            }
        }
        return nil
    }
    func firstDecodable<T: Decodable>(_ keys: [String]) -> T? {
        for key in keys {
            if let dk = DynamicKey(stringValue: key),
               let v = try? decodeIfPresent(T.self, forKey: dk) {
                return v
            }
        }
        return nil
    }
}

File: wagerproof-android-native/core/models/src/main/kotlin/com/wagerproof/core/models/EditorPick.kt
package com.wagerproof.core.models

import com.wagerproof.core.models.serialization.FallbackEnumSerializer
import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.SerializationException
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.descriptors.buildClassSerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.JsonDecoder
import kotlinx.serialization.json.JsonEncoder
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.put

/**
 * `editors_picks` row (main Supabase). Field names match the RN table
 * byte-for-byte. Decode is synthesized/non-tolerant on iOS — required fields
 * here are non-null with no default so a missing column fails the same way.
 */
@Serializable
data class EditorPick(
    val id: String,
    @SerialName("game_id") val gameId: String,
    @SerialName("game_type") val gameType: GameType,
    @SerialName("editor_id") val editorId: String,
    @SerialName("selected_bet_type") val selectedBetType: String,
    @SerialName("editors_notes") val editorsNotes: String? = null,
    @SerialName("is_published") val isPublished: Boolean,
    @SerialName("created_at") val createdAt: String,
    @SerialName("updated_at") val updatedAt: String,
    @SerialName("betslip_links") val betslipLinks: Map<String, String>? = null,
    @SerialName("pick_value") val pickValue: String? = null,
    @SerialName("best_price") val bestPrice: String? = null,
    val sportsbook: String? = null,
    val units: Double? = null,
    @SerialName("is_free_pick") val isFreePick: Boolean? = null,
    @SerialName("archived_game_data") val archivedGameData: ArchivedGameData? = null,
    @SerialName("bet_type") val betType: String? = null,
    val result: PickResult? = null,
)

/**
 * Sport keys used by `editors_picks.game_type`. Input is lowercased before
 * matching and unknown values decode to [UNKNOWN] so a stray row doesn't
 * blow up the whole feed (mirrors the Swift custom init).
 */
@Serializable(with = GameTypeSerializer::class)
enum class GameType(val raw: String) {
    NFL("nfl"),
    CFB("cfb"),
    NBA("nba"),
    NCAAB("ncaab"),
    MLB("mlb"),
    UNKNOWN("unknown");

    val displayLabel: String
        get() = when (this) {
            NFL -> "NFL"
            CFB -> "CFB"
            NBA -> "NBA"
            NCAAB -> "NCAAB"
            MLB -> "MLB"
            UNKNOWN -> "—"
        }
}

object GameTypeSerializer : FallbackEnumSerializer<GameType>(
    "GameType",
    GameType.entries.associateBy { it.raw },
    { it.raw },
    GameType.UNKNOWN,
    candidates = { listOf(it.lowercase()) },
)

/** `result` column on `editors_picks`. No fallback — unknown values throw, like Swift. */
@Serializable
enum class PickResult(val raw: String) {
    @SerialName("won") WON("won"),
    @SerialName("lost") LOST("lost"),
    @SerialName("push") PUSH("push"),
    @SerialName("pending") PENDING("pending"),
}

/**
 * Sub-doc inside `editors_picks.archived_game_data`. RN's persistence is
 * inconsistent — some keys are camelCased (`awayTeam`), others snake_cased
 * (`away_team`) — so the decoder accepts both (camelCase checked first).
 * Encoding always uses camelCase (RN's newer format).
 */
@Serializable(with = ArchivedGameDataSerializer::class)
data class ArchivedGameData(
    val awayTeam: String? = null,
    val homeTeam: String? = null,
    val awayLogo: String? = null,
    val homeLogo: String? = null,
    val gameDate: String? = null,
    val gameTime: String? = null,
    val rawGameDate: String? = null,
    val awaySpread: Double? = null,
    val homeSpread: Double? = null,
    val overLine: Double? = null,
    val awayMl: Int? = null,
    val homeMl: Int? = null,
    val awayTeamColors: TeamColors? = null,
    val homeTeamColors: TeamColors? = null,
)

object ArchivedGameDataSerializer : KSerializer<ArchivedGameData> {
    override val descriptor: SerialDescriptor = buildClassSerialDescriptor("ArchivedGameData")

    override fun deserialize(decoder: Decoder): ArchivedGameData {
        val input = decoder as? JsonDecoder
            ?: throw SerializationException("ArchivedGameData supports JSON only")
        val obj = input.decodeJsonElement() as? JsonObject ?: JsonObject(emptyMap())

        // Mirrors Swift's firstString/firstDouble/... — a wrong-typed or null
        // value under the camelCase key falls through to the snake_case key.
        fun firstString(vararg keys: String): String? = keys.firstNotNullOfOrNull { k ->
            (obj[k] as? JsonPrimitive)?.takeIf { it.isString }?.content
        }
        fun firstDouble(vararg keys: String): Double? = keys.firstNotNullOfOrNull { k ->
            (obj[k] as? JsonPrimitive)?.takeUnless { it.isString }?.doubleOrNull
        }
        fun firstInt(vararg keys: String): Int? = keys.firstNotNullOfOrNull { k ->
            (obj[k] as? JsonPrimitive)?.takeUnless { it.isString }?.intOrNull
        }
        fun firstColors(vararg keys: String): TeamColors? = keys.firstNotNullOfOrNull { k ->
            obj[k]?.let { el ->
                runCatching { input.json.decodeFromJsonElement(TeamColors.serializer(), el) }.getOrNull()
            }
        }

        return ArchivedGameData(
            awayTeam = firstString("awayTeam", "away_team"),
            homeTeam = firstString("homeTeam", "home_team"),
            awayLogo = firstString("awayLogo", "away_logo"),
            homeLogo = firstString("homeLogo", "home_logo"),
            gameDate = firstString("gameDate", "game_date"),
            gameTime = firstString("gameTime", "game_time"),
            rawGameDate = firstString("rawGameDate", "raw_game_date"),
            awaySpread = firstDouble("awaySpread", "away_spread"),
            homeSpread = firstDouble("homeSpread", "home_spread"),
            overLine = firstDouble("overLine", "over_line"),
            awayMl = firstInt("awayMl", "away_ml"),
            homeMl = firstInt("homeMl", "home_ml"),
            awayTeamColors = firstColors("awayTeamColors", "away_team_colors"),
            homeTeamColors = firstColors("homeTeamColors", "home_team_colors"),
        )
    }

    override fun serialize(encoder: Encoder, value: ArchivedGameData) {
        val output = encoder as? JsonEncoder
            ?: throw SerializationException("ArchivedGameData supports JSON only")
        val obj = buildJsonObject {
            value.awayTeam?.let { put("awayTeam", it) }
            value.homeTeam?.let { put("homeTeam", it) }
            value.awayLogo?.let { put("awayLogo", it) }
            value.homeLogo?.let { put("homeLogo", it) }
            value.gameDate?.let { put("gameDate", it) }
            value.gameTime?.let { put("gameTime", it) }
            value.rawGameDate?.let { put("rawGameDate", it) }
            value.awaySpread?.let { put("awaySpread", it) }
            value.homeSpread?.let { put("homeSpread", it) }
            value.overLine?.let { put("overLine", it) }
            value.awayMl?.let { put("awayMl", it) }
            value.homeMl?.let { put("homeMl", it) }
            value.awayTeamColors?.let {
                put("awayTeamColors", output.json.encodeToJsonElement(TeamColors.serializer(), it))
            }
            value.homeTeamColors?.let {
                put("homeTeamColors", output.json.encodeToJsonElement(TeamColors.serializer(), it))
            }
        }
        output.encodeJsonElement(obj)
    }
}

@Serializable
data class TeamColors(
    val primary: String,
    val secondary: String,
) {
    companion object {
        val DEFAULT = TeamColors(primary = "#6B7280", secondary = "#9CA3AF")
    }
}

/**
 * Per-pick game context joined from the relevant sport's input/lines table.
 * Client-assembled — not serialized.
 */
data class EditorPickGameData(
    var awayTeam: String,
    var homeTeam: String,
    var awayLogo: String? = null,
    var homeLogo: String? = null,
    var gameDate: String? = null,
    var gameTime: String? = null,
    /** ISO-ish raw date string used for sorting/filtering; `YYYY-MM-DD` or full ISO 8601. */
    var rawGameDate: String? = null,
    var awaySpread: Double? = null,
    var homeSpread: Double? = null,
    var overLine: Double? = null,
    var awayMl: Int? = null,
    var homeMl: Int? = null,
    var openingSpread: Double? = null,
    var awayTeamColors: TeamColors = TeamColors.DEFAULT,
    var homeTeamColors: TeamColors = TeamColors.DEFAULT,
)
